### DISCUSSION TITLE
Idea: Godot Wasm RL Agents

### DISCUSSION THREAD

**Discussion by gmlewis** [ashtonmeuser/godot-wasm#62] [Proposals]
First off, thank you for creating this project, @ashtonmeuser !  This is fantastic.

For background, I have found the following projects to be very inspirational:
* The Unity Machine Learning Agents Toolkit (ML-Agents): https://github.com/Unity-Technologies/ml-agents/
* Godot Reinforcement Learning (RL) Agents: https://github.com/edbeeching/godot_rl_agents/

and from the [Lunatic](https://lunatic.solutions/) project, it seems that passing strings to/from wasm is possible and common.

For years (on and off), I have been experimenting with an algorithm called "Gene Expression Programming" by Dr. Candida Ferreira that I find to be fascinating and suddenly it looks like there is a possibility to have an amazing open source playground environment that could assist with the training and execution of GEP models.

So I was thinking that Convolutional Neural Networks are well covered by the above two projects, but a platform for "general" machine-learning algorithms is not (to the best of my knowledge).

Additionally, I believe that wasm is going to be the future of computing, and I was thinking it would be wonderful to be able to use any programming language whatsoever (I currently prefer Go for the record 😄) to play around with various different machine-learning algorithms (other than CNNs).

I'm thinking that a Godot module would provide a "Component" with all the RL Agent functionality needed to point to an arbitrary wasm resource file that adheres to a simple API that could operate in two modes: training or inference. In training mode, this component would provide inputs and get outputs from the wasm module and would then control an agent within the Godot environment and provide the reward back to the wasm module. In inference mode, the wasm module would just control the agent. 
I would like there to be absolutely no C# involved (and completely rely on GDScript and Wasm) so that it would be most generally usable without the need for dotnet. (Especially after listening to this talk: https://youtu.be/tzt36EGKEZo )

TL;DR: Now for the whole point of opening up this discussion... as a Godot newbie, would you recommend that "Godot Wasm RL Agents" would be functionality that should be added directly to your "godot-wasm" repo, or do you think it would make most sense to as a separate project that simply depends upon this project?

Thanks again, as this is amazing!

**Comment by ashtonmeuser** [comment]
Thanks, @gmlewis! Really appreciate your kind words. I too believe WebAssembly will be hugely impactful.

This is a pretty large pitch you've presented but the TL;DR is helpful in summing it up with the question "Should the Godot Wasm project include a Wasm-based RL agents API?". I'd say the answer is certainly not for the following reason(s):
1. Godot Wasm is focussed on being a very low-level interface between Godot/GDScript and WebAssembly modules. Higher-level projects using Wasm in Godot should have an unassuming base upon which to build and I'm hoping Godot Wasm can be that base. I've intentionally avoided endorsing a specific use case or programming language with which to create modules.

That said, your idea is very interesting! I'd encourage you to create your own project built upon Godot Wasm and would be happy to help where I can. The simplest way to accomplish this would be either forking the project or including it as a Godot module (see [installation documentation](https://github.com/ashtonmeuser/godot-wasm/wiki/Getting-Started#installation)). An interface to train or infer from a RL model could be created on top of the basic Wasm module interface.

You mention passing strings over the host/guest barrier which is indeed possible with this library. The only stipulation is that both sides must agree on the format in which strings are to marshalled. For a concrete example using Godot's native string marshalling, take a look at the test suite ([host GDScript source](https://github.com/ashtonmeuser/godot-wasm/blob/ac9bb24ae9fbc1bceb920016b4dda2cd698a9c3a/examples/wasm-test/TestMemory.gd#L116) and [guest AssemblyScript source](https://github.com/ashtonmeuser/wasm-languages/blob/7f70a7ea98e8604ac1d915754f61973a11644ffd/AssemblyScript/assembly/memory.ts#L16)). For context, Godot marshalls strings as 4 bytes representing string length followed by the string in UTF8.

You may want to consider that model inference using Wasm will likely be slower that other implementations due to lack of access to GPU, etc. Additionally, you may want to consider splitting training and inference into separate Godot modules/addons as training is likely a dev-oriented task while inference will be done by the end user as well. Bundling training with a shipped game might be inefficient. 

Finally, you may be interested in a project I briefly played around with focussed on RL model inference in Godot using TensorFlow found [here](https://github.com/ashtonmeuser/godot-tf-inference).

**Reply by gmlewis** [reply]
Excellent!  Thank you, @ashtonmeuser, that makes sense.

**Reply by gmlewis** [reply]
I'm also curious if https://extism.org/ might be beneficial to make wasm interfaces easier.
